### ISSUE TITLE
Missing block ciphers

### ISSUE THREAD

**Issue by newpavlov** [RustCrypto/block-ciphers#1]
List of "would be nice to have" block ciphers:
- [ ] ARIA
- [ ] Camellia
- [ ] CAST5
- [ ] CAST6
- [ ] IDEA
- [ ] Kalyna
- [ ] Kuznyechik
- [ ] Magma
- [ ] RC2
- [ ] RC5
- [ ] RC6
- [ ] Serpent
- [ ] Threefish
- [ ] Twofish
- [ ] XTEA

**Comment by crypto-universe** [comment]
What about SPECK?
https://en.wikipedia.org/wiki/Speck_%28cipher%29

**Comment by newpavlov** [comment]
Missed it for some reason while compiling the list. Added Speck and Simon to it.

**Comment by k3d3** [comment]
Is Rijndael/AES being covered?

**Comment by newpavlov** [comment]
AES is present in the rust-crypto codebase, so it's already "implemented", this is why I haven't included it into this list. But it's not the easiest code to work with and better implementations exist (e.g. one in the ring), so for now it's not a highest priority for me.

**Comment by gsingh93** [comment]
I'm claiming DES, just need a bit of time to finish up Grostl over in the hashes repo before starting it.

**Comment by Trojan295** [comment]
I started to work on the RC2 cipher.

**Comment by Trojan295** [comment]
BTW. What about modes of operation for the block ciphers (CBC, OFB, etc.)? In this repo we have only the raw block ciphers. How do we progress to make them usable in different modes?

**Comment by newpavlov** [comment]
@Trojan295 
Sorry for the late answer. They will be implemented generically, though not sure if they should be placed here or in the traits repo, also I haven't yet decided on how exactly API should look like. We will probably need some kind of generic trait which will unite block ciphers under different modes of operation and stream ciphers.

**Comment by Trojan295** [comment]
I will start working on the Serpent implementation.

**Comment by labdsf** [comment]
Working on twofish, PR #7 

It is used in passwordsafe password manager, I wanted to port it to Rust but twofish package on crates.io seems to be reserved for this project and there is no implementation yet.

**Comment by dignifiedquire** [comment]
Implemented Cast5 in https://github.com/RustCrypto/block-ciphers/pull/36

**Comment by james-darkfox** [comment]
[OCB3](https://competitions.cr.yp.to/round3/ocbv11.pdf), a solid high-performance CAESAR candidate that could do with a Rust implementation.

The biggest issue with OCB is that it is patented; but Rogaway has [public free licenses](https://web.cs.ucdavis.edu/~rogaway/ocb/license.htm) available since 2013.

>  License 1 — License for Open-Source Software Implementations of OCB (Jan 9, 2013)
Under this license, you are authorized to make, use, and distribute open-source software implementations of OCB. This license terminates for you if you sue someone over their open-source software implementation of OCB claiming that you have a patent covering their implementation.
>
> License 2 — General License for Non-Military Software Implementations OCB (Jan 10, 2013).
This license does not authorize any military use of OCB. Aside from military uses, you are authorized to make, use, and distribute (1) any software implementation of OCB and (2) non-software implementations of OCB for noncommercial or research purposes. You are required to include notice of this license to users of your work so that they are aware of the prohibition against military use. This license terminates for you if you sue someone over an implementation of OCB authorized by this license claiming that you have a patent covering their implementation.

**Comment by tarcieri** [comment]
Unfortunately Rogaway's patents aren't the only ones that matter:

> Jutla (IBM)— 6,963,976, 7,093,126, and 8,107,620—and of Gligor and Donescu (VDG)—6,973,187.

https://web.cs.ucdavis.edu/~rogaway/ocb/patent-jutla-1.pdf
https://web.cs.ucdavis.edu/~rogaway/ocb/patent-jutla-2.pdf
https://web.cs.ucdavis.edu/~rogaway/ocb/patent-jutla-3.pdf
https://web.cs.ucdavis.edu/~rogaway/ocb/patent-gligor-1.pdf

**Comment by james-darkfox** [comment]
> Gligor  and  Donescu  (VDG)  and  Jutla  (IBM)  are  inventors  (owners)  on  US  patents  6,963,976,  6,973,187, 7,093,126, and 8,107,620, **all which concern AE but which may or may not apply to OCB.**

"may or may not" uh. When even the authors of the mode doesn't know. :/

**Comment by tarcieri** [comment]
In particular, Jutla 7,093,126, and 8,107,620 very much apply to OCB, IMO:

- 7,093,126: "Encryption schemes with almost free integrity awareness"
- 8,107,620: "Simple and efficient one-pass authenticated encryption scheme"

**Comment by akhilles** [comment]
@newpavlov, can `SM4` be added to the list?

**Comment by clintnetwork** [comment]
Serpent 🐍 needed!

**Comment by yerke** [comment]
@newpavlov @tarcieri I think checkbox for threefish can be checked now, since https://github.com/RustCrypto/block-ciphers/pull/5 was merged, right?

**Comment by Pure-Peace** [comment]
Rijndael - 256-bit blocks?
It seems to be similar to aes (128-bit blocks), or can you tell me how to achieve it?

**Comment by tarcieri** [comment]
AES is effectively a subset of Rijndael, so if we were to support it, it would probably make sense for it to either be part of the `aes` crate or reuse parts of it (e.g. making some of the private API public under a special feature flag)

However, it's a bit tricky because our implementation is currently heavily specialized to AES and there are multiple backends, all of which would need to be modified to support a more general Rijndael. As an example, the number of rounds varies only with the key size in AES, whereas in the more general Rijndael it varies with either/both the key size and block size.

It's something we could potentially do although I would want to be careful that we don't overcomplicate or otherwise harm the AES implementation by doing so, which might be tricky.

**Comment by sorairolake** [comment]
Implemented Camellia in #293.

**Comment by lumag** [comment]
ARIA implementation: #340

**Comment by sorairolake** [comment]
@newpavlov @tarcieri Why isn't `speck-cipher` crate (Speck) published yet?

**Comment by tarcieri** [comment]
@sorairolake unfortunately only @newpavlov currently has access to publish it.

I would suggest we publish a v0.0.1 based on 8b1499cf9b9a478caee0ee7f9874fe195ae60573, since the current `master` is on the newest prerelease `cipher` crate.

**Comment by newpavlov** [comment]
Oh, we have indeed forgot to publish the Speck crate (I think we were hoping for potential transfer of the `speck` name). I've added the block ciphers group to owners of the `speck-cipher` crate.

**Comment by tarcieri** [comment]
It's released: https://crates.io/crates/speck-cipher/0.0.1

**Comment by sorairolake** [comment]
I think it would be good to add an implementation of the [CLEFIA](https://en.wikipedia.org/wiki/CLEFIA) block cipher as described in [RFC 6114](https://datatracker.ietf.org/doc/html/rfc6114) to this project in the future. This block cipher is the one of three 128-bit block ciphers listed on the Candidate Recommended Ciphers List by the Japanese Government's [CRYPTREC](https://en.wikipedia.org/wiki/CRYPTREC) as of April 2026.

However, as of 2012, this block cipher was protected by intellectual property rights and could not be implemented royalty-free.[^1][^2] In other words, at least at that time, it was not licensed under a  royalty-free, reasonable and non-discriminatory license to all implementers. I'm unaware of the current intellectual property rights status of it.

I previously implemented the [Camellia](https://en.wikipedia.org/wiki/Camellia_(cipher)) block cipher as described in [RFC 3713](https://datatracker.ietf.org/doc/html/rfc3713) and the [KCipher-2](https://en.wikipedia.org/wiki/KCipher-2) stream cipher as described in [RFC 7008](https://datatracker.ietf.org/doc/html/rfc7008) in pure Rust. Before implementing these ciphers, I confirmed that they were licensed under royalty-free, reasonable and non-discriminatory licenses to all implementers.[^3][^4][^5][^6]

Therefore, I believe that this block cipher can only be implemented as FOSS after it has been definitively confirmed that all intellectual property rights to it have expired.

[^1]: [Sony Corporation's Statement about IPR related to RFC 6114](https://datatracker.ietf.org/ipr/1586/)
[^2]: [知的財産権の使用の権利に係る確認書](https://www.cryptrec.go.jp/license/PDF/22_CLEFIA_license.pdf)
[^3]: [Intellectual Property Information](https://info.isl.ntt.co.jp/crypt/eng/info/chiteki.html)
[^4]: [知的財産権の使用の権利に係る確認書](https://www.cryptrec.go.jp/license/PDF/06_Camellia_license.pdf)
[^5]: [KDDI Corporation's Statement about IPR related to draft-kiyomoto-kcipher2-06](https://datatracker.ietf.org/ipr/1916/)
[^6]: [知的財産権の使用の権利に係る確認書](https://www.cryptrec.go.jp/license/PDF/21_KCipher-2_license.pdf)

**Comment by tarcieri** [comment]
@sorairolake we probably want to avoid any constructions with IPR concerns

**Comment by randombit** [comment]
Any interest in [SHACAL2](https://en.wikipedia.org/wiki/SHACAL)? Approved by NESSIE and the best known attack breaks 44 of 64 rounds using 2^223 related key chosen ciphertexts with 2^497 complexity. Any realistic attack on it probably implies a break of SHA-256. Admittedly a bit obscure but has many good points - 256 bit block (so birthday problem type issues that can apply to 128-bit ciphers are a complete non-issue), ARX style ciphers are trivially constant time and easy to run as SIMD, can use SHA-256 instructions (SHA-NI/ARMv8) for the encryption direction, and even a baseline scalar version is quite fast.
